packages/react-dom/src/events/DOMModernPluginEventSystem.js

```diff
@@ -165,6 +167,57 @@ export const reactScopeListenerStore: WeakMap<
   >,
 > = new PossiblyWeakMap();
 
+function executeDispatch(
+  event: ReactSyntheticEvent,
+  listener: Function,
+  currentTarget: EventTarget,
+): void {
+  const type = event.type || 'unknown-event';
+  event.currentTarget = currentTarget;
+  invokeGuardedCallbackAndCatchFirstError(type, listener, undefined, event);
+  event.currentTarget = null;
+}
+
+function executeDispatchesInOrder(event: ReactSyntheticEvent): void {
+  // TODO we should remove _dispatchListeners and _dispatchInstances at some point.
+  const dispatchListeners = event._dispatchListeners;
+  const dispatchInstances = event._dispatchInstances;
+  let previousInstance;
+
+  if (dispatchListeners !== null && dispatchInstances !== null) {
+    for (let i = 0; i < dispatchListeners.length; i++) {
+      const dispatchInstance = dispatchInstances[i];
+      const dispatchListener = dispatchListeners[i];
+      let instance = dispatchInstance;
+      let currentTarget;
+
+      // dispatchInstance can come in two forms:
+      // - as a Fiber instance
+      // - as an object { instance: Fiber | null, container: EventTarget }
+      // The second form is used for useEvent types
+      if (dispatchInstance.container !== undefined) {
+        instance = ((dispatchInstance: any): DispatchInstance).instance;
+        currentTarget = ((dispatchInstance: any): DispatchInstance).container;
+      } else {
+        currentTarget = getNodeFromInstance(((dispatchInstance: any): Fiber));
+      }
+      // We check if the instance was the same as the last one,
+      // if it was, then we're still on the same instance thus
+      // propagation should not stop. If we add support for
+      // stopImmediatePropagation at some point, then we'll
+      // need to handle that case here differently.
+      if (instance !== previousInstance && event.isPropagationStopped()) {
```

Comment:
I'm still not sure I understand this piece of logic. We check for propagation for the first `useEvent` OK. Then we have another `useEvent` on the same node. Let's say this one calls `stopPropagation`. Would we ignore it? Why?

Reply:
We'd ignore it as `stopPropagation` works like that on the DOM today. We'd need to add `stopImmediatePropagation` that works like how you probably want, where we can stop propagation even on the same instance.

Reply:
Hmm. I'm not sure I understand what you mean by DOM working like this.

Can you provide an example?

>We'd need to add stopImmediatePropagation that works like how you probably want

I am not referring to one `useEvent` preventing another `useEvent`.

I am talking about the second `useEvent` being able to prevent *further* propagation up the tree. Does that work?

Reply:
> I am talking about the second useEvent being able to prevent further propagation up the tree. Does that work?

Yes it does, by design.

> Hmm. I'm not sure I understand what you mean by DOM working like this.

I mean that it works in same way propagation works on the DOM. We move a pointer up and down a single-list (which is section of the node tree) and each level of the tree is a propagation layer. If something stops propagation, we don't exit processing of the list until we move the pointer to the next level.

Reply:
>Yes it does, by design.

Do you mind adding a test for this?

Reply:
That's what the test on this PR should be testing, unless we're talking about different things.

Reply:
OK I get it now. What you're saying is that by the time we reach the next node, we will run this check again, and so if it has been stopped by then, we won't dispatch for it.